src/main/scala/org/broadinstitute/dsde/workbench/leonardo/model/LeoAuthProvider.scala

```diff
@@ -24,26 +26,56 @@ object NotebookClusterActions {
 
 }
 
-abstract class LeoAuthProvider(authConfig: Config) {
+abstract class LeoAuthProvider(authConfig: Config, serviceAccountProvider: ServiceAccountProvider) {
   /**
-    * @param userInfo The user in question
+    * @param userEmail The user in question
     * @param action The project-level action (above) the user is requesting
     * @param googleProject The Google project to check in
     * @return If the given user has permissions in this project to perform the specified action.
     */
-  def hasProjectPermission(userInfo: UserInfo, action: ProjectActions.ProjectAction, googleProject: String): Future[Boolean]
+  def hasProjectPermission(userEmail: WorkbenchEmail, action: ProjectActions.ProjectAction, googleProject: GoogleProject)(implicit executionContext: ExecutionContext): Future[Boolean]
+
+  /**
+    * When listing clusters, Leo will perform a GROUP BY on google projects and call this function once per google project.
+    * If you have an implementation such that users, even in some cases, can see all clusters in a google project, overriding
+    * this function may lead to significant performance improvements.
+    * For any projects where this function call returns Future.successful(false), Leo will then call hasNotebookClusterPermission
+    * for every cluster in that project, passing in action = GetClusterStatus.
+    *
+    * @param userEmail The user in question
+    * @param googleProject A Google project
+    * @return If the given user can see all clusters in this project
+    */
+  def canSeeAllClustersInProject(userEmail: WorkbenchEmail, googleProject: GoogleProject)(implicit executionContext: ExecutionContext): Future[Boolean] = {
+    Future.successful(false)
+  }
+
+  /**
+    * When listing clusters, Leo will perform a GROUP BY on google projects and call this function once per google project.
+    * If you have an implementation such that users, even in some cases, can see all clusters in a google project, overriding
+    * this function may lead to significant performance improvements.
+    * For any projects where this function call returns Future.successful(false), Leo will then call hasNotebookClusterPermission
+    * for every cluster in that project, passing in action = GetClusterStatus.
+    *
+    * @param userEmail The user in question
+    * @param googleProject A Google project
+    * @return If the given user can see all clusters in this project
+    */
+  def canSeeAllClustersInProject(userEmail: WorkbenchEmail, googleProject: String): Future[Boolean] = {
```

Comment:
Not sure we need to overload this method to take a `String` instead of a `GoogleProject` if we don't do it anywhere else. Maybe just have the `GoogleProject` version (to be consistent), and we can fix everything at the same time if we decide to `String`ify things?

Also this version doesn't take an `ExecutionContext`.